File: Array/Merge Sort Array/mergeSortArray.js
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
function mergeSortArray(nums1, m, nums2, n) {
    let i = m - 1; // Index in nums1 for elements from nums1
    let j = n - 1; // Index in nums2
    let k = m + n - 1; // Index in nums1 for the merged result

    // While there are elements in both arrays
    while (i >= 0 && j >= 0) {
        // Compare elements from the end of both arrays and place the larger one in nums1
        if (nums1[i] > nums2[j]) {
            nums1[k] = nums1[i];
            i--;
        } else {
            nums1[k] = nums2[j];
            j--;
        }
        k--;
    }

    // If there are remaining elements in nums2, copy them to nums1
    while (j >= 0) {
        nums1[k] = nums2[j];
        j--;
        k--;
    }

    console.log(nums1)
};

console.log(merge([1,2,3,0,0,0], 3, [2,5,6], 3));//[1,2,2,3,5,6]
console.log(merge([1],1, [], 0));//[1]
console.log(merge([0], 0, [1], 1));//[1]

File: 88 - Merge Sort Array/MergeSortArray.java
import java.util.Arrays;

class MergeSortArray {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        // Start from the end of both arrays
        int i = m - 1; // Last element in nums1's initial part
        int j = n - 1; // Last element in nums2
        int k = m + n - 1; // Last element in nums1's combined size

        // Merge in reverse order
        while (i >= 0 && j >= 0) {
            if (nums1[i] > nums2[j]) {
                nums1[k--] = nums1[i--];
            } else {
                nums1[k--] = nums2[j--];
            }
        }

        // If there are remaining elements in nums2, add them
        while (j >= 0) {
            nums1[k--] = nums2[j--];
        }
    }

    public static void main(String[] args) {
        MergeSortArray solution = new MergeSortArray();

        // Test Case 1
        int[] nums1_1 = {1, 2, 3, 0, 0, 0};
        int m1 = 3;
        int[] nums2_1 = {2, 5, 6};
        int n1 = 3;
        solution.merge(nums1_1, m1, nums2_1, n1);
        System.out.println("Merged array (Test Case 1): " + Arrays.toString(nums1_1)); // Output: [1, 2, 2, 3, 5, 6]

        // Test Case 2
        int[] nums1_2 = {1};
        int m2 = 1;
        int[] nums2_2 = {};
        int n2 = 0;
        solution.merge(nums1_2, m2, nums2_2, n2);
        System.out.println("Merged array (Test Case 2): " + Arrays.toString(nums1_2)); // Output: [1]

        // Test Case 3
        int[] nums1_3 = {0};
        int m3 = 0;
        int[] nums2_3 = {1};
        int n3 = 1;
        solution.merge(nums1_3, m3, nums2_3, n3);
        System.out.println("Merged array (Test Case 3): " + Arrays.toString(nums1_3)); // Output: [1]
    }
}
